service/storage_service.cc

```diff
@@ -699,27 +699,58 @@ future<> storage_service::topology_transition() {
 }
 
 future<> storage_service::merge_topology_snapshot(raft_topology_snapshot snp) {
-    std::vector<mutation> muts;
-    muts.reserve(snp.topology_mutations.size() + snp.cdc_generation_mutations.size() + snp.topology_requests_mutations.size());
-    {
-        auto s = _db.local().find_schema(db::system_keyspace::NAME, db::system_keyspace::TOPOLOGY);
-        boost::transform(snp.topology_mutations, std::back_inserter(muts), [s] (const canonical_mutation& m) {
-            return m.to_mutation(s);
-        });
-    }
-    if (snp.cdc_generation_mutations.size() > 0) {
+    if (!snp.cdc_generation_mutations.empty()) {
         auto s = _db.local().find_schema(db::system_keyspace::NAME, db::system_keyspace::CDC_GENERATIONS_V3);
-        boost::transform(snp.cdc_generation_mutations, std::back_inserter(muts), [s] (const canonical_mutation& m) {
-            return m.to_mutation(s);
-        });
+
+        // Split big mutations into smaller ones, prepare frozen_muts_to_apply
+        std::vector<frozen_mutation> frozen_muts_to_apply;
+        {
+            std::vector<mutation> muts_to_apply;
+            muts_to_apply.reserve(snp.cdc_generation_mutations.size());
+            const auto max_size = _db.local().commitlog()->max_record_size() / 2;
+            for (const auto& m: snp.cdc_generation_mutations) {
+                auto mut = m.to_mutation(s);
+                if (m.representation().size() <= max_size) {
+                    muts_to_apply.push_back(std::move(mut));
+                } else {
+                    co_await split_mutation(std::move(mut), muts_to_apply, max_size);
+                }
+            }
+            frozen_muts_to_apply = freeze(muts_to_apply);
+        }
+
+        // Apply non-atomically so as not to hit the commitlog size limit.
+        // The cdc_generations_v3 data is not used in any way until
+        // it's referenced from the topology table.
+        // By applying the cdc_generations_v3 mutations before topology mutations
+        // we ensure that the lack of atomicity isn't a problem here.
+        std::vector<future<>> futures;
+        futures.reserve(frozen_muts_to_apply.size());
+        for (const auto& m: frozen_muts_to_apply) {
+            futures.push_back(_db.local().apply(s, m, {}, db::commitlog::force_sync::no, db::no_timeout));
+        }
+        co_await when_all_succeed(futures.begin(), futures.end()).discard_result();
```

Comment:
Aren't you reimplementing `parallel_for_each`?
(Also perhaps we should limit concurrency with `max_concurrent_for_each`)